### ISSUE TITLE
Sentry bundle breaks doctrine migrations.

### ISSUE THREAD

**Issue by vuryss** [getsentry/sentry-symfony#488]
### Environment

Latest symfony 5.2 + doctrine migrations 3.1 + sentry-symfony 4.1

### Steps to Reproduce

Having the above combination completely breaks the Doctrine migrations unless we add to each migrations a method indicating that the migration is not transactional.
Doctrine had this issue before but they fixed it with this check:

https://github.com/doctrine/migrations/blob/96c64fa17477e503da8245f14541d70d7ae76091/lib/Doctrine/Migrations/Tools/TransactionHelper.php#L39

However, sentry decorates the connection with:

`Sentry\SentryBundle\Tracing\Doctrine\DBAL\TracingDriverConnection`  

which is obviously not a PDO instance, so the aforementioned check fails and we get a nasty error:

```
In TracingDriverConnection.php line 163:
                                  
  [PDOException]                  
  There is no active transaction  
```                                  

### Expected Result

Well obviously I expect migrations to work with and without the sentry bundle :)

### Actual Result

```
In TracingDriverConnection.php line 163:
                                  
  [PDOException]                  
  There is no active transaction  
```

**Comment by ste93cry** [comment]
I think there is nothing we can do here for the moment, since the check done in the Doctrine repo is clearly intended to work only for PDO-based drivers. @greg0ire is there any reason for that check to be like that? Maybe we could make the `inTransaction` method part of the `Doctrine\DBAL\Driver\Connection` interface so that the check in `TransactionHelper` does not have to care whether or not the connection is a PDO connection, but it just has to ask whether the transaction is active or not. Each driver connection can then report back either `true` or `false` based on specific conditions. Otherwise, any attempt to wrap the driver's connection will cause the same issue we're having here

**Comment by greg0ire** [comment]
I think the reason was we needed a fix that would not require bumping our dependency on the DBAL, but you are right about `inTransaction()` possibly being the right API: https://github.com/doctrine/dbal/pull/4544#issuecomment-799596645

**Comment by ste93cry** [comment]
Understood. Do you think it would be feasible to implement this API in both version 2.x and 3.x? And do you think there would be anyone in the Doctrine team interested in doing this?

**Comment by greg0ire** [comment]
> Do you think it would be feasible to implement this API in both version 2.x and 3.x?

Feasible, sure, but I think you and I both know it will not be accepted: https://github.com/doctrine/dbal/issues/4495

> And do you think there would be anyone in the Doctrine team interested in doing this?

Maybe Sergei will? I will open a new issue so that we can investigate this is indeed doable.

**Comment by ste93cry** [comment]
>Feasible, sure, but I think you and I both know it will not be accepted

I completely forgot we already had this discussion for another thing 😅 I'm not sure what falls under the "upgrade path" category mentioned in that PR, but I see unfeasible to stick with that reason until the ORM does not support DBAL `3.x`

**Comment by greg0ire** [comment]
Issue opened at https://github.com/doctrine/dbal/issues/4616

> but I see unfeasible to stick with that reason until the ORM does not support DBAL 3.x

let's see how things look once it's implemented in 3.x? It will still probably be useful in the future.

**Comment by ste93cry** [comment]
Yea, having it in `3.x` would already be a great starting point. However,  our bundle and any other code that wraps the driver connection is not working properly with `doctrine/migrations` as of now and since I bet it will take a bit until things get implemented in the DBAL maybe we could implement a workaround until that time?

**Comment by greg0ire** [comment]
I suppose you could copy/paste `TransactionHelper` in this bundle?

**Comment by greg0ire** [comment]
Wait actually it will? You could use it here: https://github.com/getsentry/sentry-symfony/blob/107f60cd87073a25b960a2ce8483e0b5cbb46d41/src/Tracing/Doctrine/DBAL/TracingDriverConnection.php#L153

**Comment by ste93cry** [comment]
Yes, however I'm wondering if it would be safe to do such change because we would change the behaviour of the `commit()` method of the connection, wouldn't we?

**Comment by greg0ire** [comment]
Ah right you want to do that only in the context of migrations somehow, is that the issue?

**Comment by ste93cry** [comment]
>Ah right you want to do that only in the context of migrations somehow, is that the issue?

Yes. In reality, the only thing we care about is that with or without this bundle things must work the same, of course. Instrumentation of the DBAL for tracing purposes should be transparent to the user and should not affect the behaviour of the application. The issue at the moment appears in the context of migrations, but to be honest I didn't test what happens if I try to execute a DDL SQL statement using the DBAL directly. My concern is that while the migrations, thanks to the usage of that helper, silently "swallow the error" (or don't make it happen), the DBAL instead would throw the exception due to the fact that there is no similar mechanism in that package. As a consequence of this, we have two distinct behaviour, and if I use the helper inside the driver connection then I'm changing the behaviour not only in the context of migrations but also of all the rest of the usages of the DBAL. Does it makes sense?

>You could do that by inspecting the stack trace but ewww…

Yes, that's a possibility, although since the driver connection is a critical code path I think it's better to not do these kind of operations there that are costly and may affect performances more than what we already do for logging the data for the distributed tracing feature.

>even if you had inTransaction at your disposal, it would change that particular issue, would it

What I was thinking is that in the `doctrine/migrations` package the condition could be changed to check whether the object instance implements the `inTransaction` method. That would basically be a poor-man version of the check based on the interface, until we have it at least, and would allow us here to add such method to the connection class which in turn will be used to determine whether the transaction is active. If in the future the DBAL will formalize the need for this method then such method would then be officially used not only in the context of the migrations but also from the DBAL itself and this fact would be completely independent from the usage of this bundle

**Comment by greg0ire** [comment]
> What I was thinking is that in the doctrine/migrations package the condition could be changed to check whether the object instance implements the inTransaction method. 

Sounds good. Can you send a PR?

**Comment by ste93cry** [comment]
Sure, I will do it asap 👍

**Comment by ndench** [comment]
Downgrading sentry-symfony to 4.0.3 has temporarily fixed this for me.

**Comment by JBx0** [comment]
Downgrading to :
`"sentry/sentry-symfony": "<4.0.9",`

works

**Comment by ste93cry** [comment]
You can also set `sentry.tracing.dbal.enabled: false` or `sentry.tracing.enabled: false` to disable tracing for the DBAL or completely and avoid downgrading

**Comment by JBx0** [comment]
> You can also set `sentry.tracing.dbal.enabled: false` or `sentry.tracing.enabled: false` to disable tracing for the DBAL or completely and avoid downgrading

sure thing if you don't use tracing, it makes sense

**Comment by BenMorel** [comment]
Same issue here. Looking forward to a fix! :+1:

**Comment by lukasluecke** [comment]
Would really like to use the tracing in general, is there an easy way to just disable the tracing programatically for the fixtures only?

**Comment by Jean85** [comment]
For the record, this should fix it downstream: https://github.com/theofidry/AliceDataFixtures/pull/174

**Comment by lkm** [comment]
Does anyone know what happening with this? Is there a workaround that doesn't involve sacrificing doctrine tracing?

**Comment by lkm** [comment]
Right so I missed the part on the doctrine/migrations thread about adding `isTransactional() { return false; }` to migrations, that fixed my immediate problem

**Comment by fritzmg** [comment]
This also breaks database migrations in Contao. Due to the decoration of the database connection, doctrine can no longer extract the correct database server version - which will then lead to wrong information about the platform and thus wrong schema diffs in `Doctrine\DBAL\Schema\Comparator::compare`.

Specifically `Doctrine\DBAL\Connection::getServerVersion` will always return `null` with the current version of `symfony/sentry-symfony` installed, because [this condition](https://github.com/doctrine/dbal/blob/cbd8cd8e41b0bd60730e8dfc1487658b4e5b8cdf/lib/Doctrine/DBAL/Connection.php#L517-L520) will always evaluate to `false`, since the decorated connection does not implement `ServerInfoAwareConnection` anymore (while the regular connection would).

**Comment by ste93cry** [comment]
>forward the call to the decorated connection accordingly (or return `null` otherwise)

I tried to do this yesterday, just to find out that the return type is not marked as nullable according to the DocBlock. Instead, apparently, once you implement the interface you have to either return a value or throw an exception, which may cause issues because only specific drivers implement it and therefore are able to give back the server version. As a consequence of this, I started working on yet another decorator which implements the ServerInfoAwareConnection interface and that will wrap the instrumented connection if the original connection was implementing the aforementioned interface. Work is not finished yet, will try to find some time in the next couple of days to see if it can work

**Comment by BenMorel** [comment]
I can confirm that we can now use the new latest Sentry version by adding this line to the doctrine_migrations config file:

```yaml
doctrine_migrations:
    transactional: false
```

**Comment by radomeit** [comment]
`doctrine_migrations:
    transactional: false`

did not work for us, we had to disable DB SQL Logging via
`tracing:
    dbal: 
        enabled: false`

in our sentry.yaml to get the migrations working again.

Otherwise we'll get the **"PDOException There is no active transaction"** when running doctrine migrations migrate.

Transactional is set to false in doctrine_migrations.yaml.

We are using
sentry/sentry-symfony 4.2.9
doctrine/doctrine-bundle 2.6.3
doctrine/doctrine-migrations-bundle 3.2.2

So this still seems to be a problem.
